src/main/java/teammates/logic/core/DataBundleLogic.java

```diff
@@ -109,18 +110,249 @@ public void persistDataBundle(DataBundle dataBundle) throws InvalidParametersExc
         processResponsesAndPopulateMap(responses, sessionResponsesMap);
         processSessionsAndUpdateRespondents(sessions, courseInstructorsMap, sessionQuestionsMap, sessionResponsesMap);
 
-        accountsDb.putEntities(googleIdAccountMap.values());
-        profilesDb.putEntities(profiles);
-        coursesDb.putEntities(courses);
-        instructorsDb.putEntities(instructors);
-        studentsDb.putEntities(students);
-        fbDb.putEntities(sessions);
+        List<AccountAttributes> newAccounts = accountsDb.putEntities(googleIdAccountMap.values());
+
+        List<StudentProfileAttributes> newProfiles = profilesDb.putEntities(profiles);
+        List<CourseAttributes> newCourses = coursesDb.putEntities(courses);
+        List<InstructorAttributes> newInstructors = instructorsDb.putEntities(instructors);
+        List<StudentAttributes> newStudents = studentsDb.putEntities(students);
+        List<FeedbackSessionAttributes> newFeedbackSessions = fbDb.putEntities(sessions);
 
         List<FeedbackQuestionAttributes> createdQuestions = fqDb.putEntities(questions);
         injectRealIds(responses, responseComments, createdQuestions);
 
-        frDb.putEntities(responses);
-        fcDb.putEntities(responseComments);
+        List<FeedbackResponseAttributes> newFeedbackResponses = frDb.putEntities(responses);
+        List<FeedbackResponseCommentAttributes> newFeedbackResponseComments = fcDb.putEntities(responseComments);
+
+        updateDataBundleAccounts(dataBundle, newAccounts);
+
+        updateDataBundleProfiles(dataBundle, newProfiles);
+        updateDataBundleCourses(dataBundle, newCourses);
+        updateDataBundleInstructors(dataBundle, newInstructors);
+
+        updateDataBundleStudents(dataBundle, newStudents);
+        updateDataBundleFeedbackSessions(dataBundle, newFeedbackSessions);
+        updateDataBundleFeedbackQuestions(dataBundle, createdQuestions);
+        updateDataBundleFeedbackResponses(dataBundle, newFeedbackResponses);
+        updateDataBundleFeedbackResponseComments(dataBundle, newFeedbackResponseComments);
+
+        return dataBundle;
+
+    }
+
+    private void updateDataBundleAccounts(DataBundle dataBundle, List<AccountAttributes> newAccounts) {
```

Comment:
seems the logic can be simplified here?
Is "adding all accounts from 'newAccounts' that already exist in oldAccounts?"

Reply:
It is mapping back the accounts in newAccounts back to their original keys, by finding an corresponding account in the oldAccounts (which is the account before being added into the server)

The oldAccounts contain the Map<string, AccountAttributes> which contains the attributes before they were added into the server. NewAccounts contain a List of AccountAttributes, that contain the same objects as oldAccounts, but taken directly from the database after being added, which means that they have updated values that their respective counterparts in oldAccounts do not have.

Since dataBundle uses Map to store the respective attributes, when we need to update the Attributes from the new ones directly from the database, we need to ensure that they get mapped back to the correct Key. This is because the external classes that uses databundle uses the keys to retrieve the attributes, and having the wrong attributes mapped to wrong keys will result in problems for those external classes.

The first problem is that when the values are passed into the database, key-value mapping is lost and we only have the list of attributes to work on. The second issue is that the old and new value aren't exactly equal, because the new attributes have some variables that are generated by the server (createdAt, and feedbackQuestionId are such variables). The third issue is that there seems to be no requirement for uniqueness for the attribute's variables, which means that there is no single variable that can be used to link the new attributes to the old attribute.

I ended up with writing this n + 1 query to ensure that the updated attributes gets mapped back to the correct key, by comparing all possible relevant variables with the old values, that does not get changed by the database. I'm definitely open to using another approach if a better one exists.

Reply:
In this case if you want to check if one entry in oldAccounts exists in newAccounts, probably hasing is a better idea? Currently the double nested loop seems to be not effiecit (O(n^2))

Reply:
Updated logic to use hashing